Sources/DeepLearning/DifferentialOperators.swift

```diff
@@ -0,0 +1,176 @@
+// ===------------------------------------------------------------------------------------------===//
+
+// Value with gradient
+
+@inlinable
+public func valueWithGradient<T, R>(
+    at x: T,
+    in f: @differentiable (T) -> Tensor<R>
+) -> (value: Tensor<R>, gradient: T.CotangentVector)
+where T: Differentiable, R: TensorFlowFloatingPoint {
+    let (y, pullback) = valueWithPullback(at: x, in: f)
+    return (y, pullback(Tensor<R>(1)))
+}
+
+@inlinable
+public func valueWithGradient<T, U, R>(
+    at x: T,
+    _ y: U,
+    in f: @differentiable (T, U) -> Tensor<R>
+) -> (value: Tensor<R>, gradient: (T.CotangentVector, U.CotangentVector))
+    where T: Differentiable, U: Differentiable, R: TensorFlowFloatingPoint {
+    let (y, pullback) = valueWithPullback(at: x, y, in: f)
+    return (y, pullback(Tensor<R>(1)))
+}
+
+// @inlinable
+// public func valueWithGradient<T, U, V, R>(
+//     at x: T,
+//     _ y: U,
+//     _ z: V,
+//     in f: @differentiable (T, U, V) -> Tensor<R>
+// ) -> (value: Tensor<R>, gradient: (T.CotangentVector, U.CotangentVector, V.CotangentVector))
+//   where T: Differentiable, U: Differentiable, V: Differentiable, R: TensorFlowFloatingPoint {
+//   let (y, pullback) = valueWithPullback(at: x, y, z, in: f)
+//   return (y, pullback(Tensor<R>(1)))
+// }
+
+// Value with gradient (curried)
+
+@inlinable
+public func valueWithGradient<T, R>(
+    of f: @escaping @differentiable (T) -> Tensor<R>
+) -> (T) -> (value: Tensor<R>, gradient: T.CotangentVector)
+    where T: Differentiable, R: TensorFlowFloatingPoint {
+    return { x in valueWithGradient(at: x, in: f) }
+}
+
+@inlinable
+public func valueWithGradient<T, U, R>(
+    of f: @escaping @differentiable (T, U) -> Tensor<R>
+) -> (T, U) -> (value: Tensor<R>, gradient: (T.CotangentVector, U.CotangentVector))
+  where T: Differentiable, U: Differentiable, R: TensorFlowFloatingPoint {
+  return { x, y in valueWithGradient(at: x, y, in: f) }
+}
+
+// @inlinable
+// public func valueWithGradient<T, U, V, R>(
+//     of f: @escaping @differentiable (T, U, V) -> Tensor<R>
+// ) -> (T, U, V) -> (
+//     value: Tensor<R>,
+//     gradient: (T.CotangentVector, U.CotangentVector, V.CotangentVector))
+//     where T: Differentiable, U: Differentiable, V: Differentiable, R: TensorFlowFloatingPoint {
+//   return { x, y, z in valueWithGradient(at: x, y, z, in: f) }
+// }
+
+// Gradient
+
+@inlinable
+public func gradient<T, R>(
+    at x: T,
+    in f: @differentiable (T) -> Tensor<R>
+) -> T.CotangentVector where T: Differentiable, R: TensorFlowFloatingPoint {
+    return pullback(at: x, in: f)(Tensor<R>(1))
+}
+
+@inlinable
+public func gradient<T, U, R>(
+    at x: T,
+    _ y: U,
+    in f: @differentiable (T, U) -> Tensor<R>
+) -> (T.CotangentVector, U.CotangentVector)
+    where T: Differentiable, U: Differentiable, R: TensorFlowFloatingPoint {
+    return pullback(at: x, y, in: f)(Tensor<R>(1))
+}
+
+// @inlinable
+// public func gradient<T, U, V, R>(
```

Comment:
Any reasons this is commented out?

Reply:
Sorry I forgot about this earlier. All the functions in this file that have 4 generic type arguments result in the following compiler crash:
```
1.      While verifying SIL function "@$s12DeepLearning17valueWithGradient2at__2in14TensorFlowCore0H0Vyq1_G0C0_15CotangentVectorQz_AJQy_AJQy0_t8gradienttx_q_q0_AHx_q_q0_tXEts14DifferentiableRzsAOR_sAOR0_AE0hI13FloatingPointR1_r2_lF".
 for <<debugloc at "<compiler-generated>":0:0>>0  swift                    0x0000000109049675 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x0000000109048915 llvm::sys::RunSignalHandlers() + 85
2  swift                    0x0000000109049c62 SignalHandler(int) + 258
3  libsystem_platform.dylib 0x00007fff60157b5d _sigtramp + 29
4  swift                    0x000000010a377052 cmark_strbuf__initbuf + 133885
5  libsystem_c.dylib        0x00007fff600176a6 abort + 127
6  libsystem_c.dylib        0x00007fff5ffe020d basename_r + 0
7  swift                    0x000000010626a268 swift::AutoDiffFunctionInst::getAssociatedFunctionPair(unsigned int) const + 136
8  swift                    0x00000001062d4228 swift::SILInstructionVisitor<(anonymous namespace)::SILVerifier, void>::visit(swift::SILInstruction*) + 47832
9  swift                    0x00000001062c73ec (anonymous namespace)::SILVerifier::visitSILBasicBlock(swift::SILBasicBlock*) + 1196
10 swift                    0x00000001062c15b7 swift::SILFunction::verify(bool) const + 8471
11 swift                    0x00000001062c4553 swift::SILModule::verify() const + 211
12 swift                    0x000000010537fadf performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 10895
13 swift                    0x000000010537bfe8 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3016
14 swift                    0x000000010532dd12 main + 690
15 libdyld.dylib            0x00007fff5ff723d5 start + 1
```
I temporarily commented out as they're not super critical but forgot to report the bug.

Reply:
Interesting. I suspect it's a small typo somewhere. It's fine to remove it since no one seems to be using the ternary version.

Reply:
Sounds good. Is it ok to keep them commented out or should I completely remove them for now?

Reply:
Commenting out is fine.

Reply:
Ok cool. In this case then this PR should be ready along with apple/swift#24261 and TensorFlow/swift-bindings#26 as everything compiles and all tests pass on my machine.